### ISSUE TITLE
Option request: LACT systemd service

### ISSUE THREAD

**Issue by viscountexx** [NixOS/nixpkgs#317544]
**Project description**

LACT, added to nixpkgs as of #237380, has no NixOS option to enable its systemd service. LACT's daemon mode is *required* to run in the background (contrary to what the original issue maker suggested), so this seems like a big oversight. I used the package request template here as it's the closest to the issue at hand.

**Metadata**

* homepage URL: https://github.com/ilya-zlobintsev/LACT
* source URL: https://github.com/ilya-zlobintsev/LACT
* license: mit
* platforms: linux

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by colemickens** [comment]
Note someone did this in their own config: https://github.com/paschoal/dotfiles/blob/5728450b8d4390776b4d6f0506fc980f95e67c7c/hardware/radeon/default.nix

Note, if upstream package includes a unit, it would be better to just include the package in the right systemd option instead.

**Comment by colemickens** [comment]
Sent a draft PR for this, using the upstream systemd file.

**Comment by mksafavi** [comment]
Is adding a new service module necessary? I checked the PR. as it doesn't have any new options, I think just installing the unit from the `lact` module accomplishes the same thing. 

I have this on my `configurations.nix` file:

```
  environment.systemPackages = with pkgs; [                                                                                                                                            
    lact                                                                                                                                                                                                                                                                     
  ];                                                                                                                                                                                                                                                                                                                                           
  systemd.packages = with pkgs; [ lact ];                                                                                                                                                    
  systemd.services.lact.enable = true; 
```
afterward that you should have `systemctl status lactd.service`

**Comment by mksafavi** [comment]
> **Project description**
> 
> LACT, added to nixpkgs as of #237380, has no NixOS option to enable its systemd service. LACT's daemon mode is _required_ to run in the background (contrary to what the original issue maker suggested), so this seems like a big oversight. I used the package request template here as it's the closest to the issue at hand.

I agree. it's confusing. For now I added a section to the [amd gpu wiki](https://wiki.nixos.org/wiki/AMD_GPU#GUI_tools)

**Comment by viscountexx** [comment]
> Is adding a new service module necessary? I checked the PR. as it doesn't have any new options, I think just installing the unit from the `lact` module accomplishes the same thing.
> 
> I have this on my `configurations.nix` file:
> 
> ```
>   environment.systemPackages = with pkgs; [                                                                                                                                            
>     lact                                                                                                                                                                                                                                                                     
>   ];                                                                                                                                                                                                                                                                                                                                           
>   systemd.packages = with pkgs; [ lact ];                                                                                                                                                    
>   systemd.services.lact.enable = true; 
> ```
> 
> afterward you should have `systemctl status lactd.service`

Oh. I uh, I've been using NixOS for ages and just assumed these services were not in pkgs at all because it was extremely unclear to me.

Huh.

Well, thank you, that clears up an issue I've had for a long time.

**Comment by mksafavi** [comment]
> Edit: Tiny note, but I think the systemd service packaged with LACT is `lactd.service`, so that last line should be `systemd.services.lactd.enable = true` - Otherwise, yep, works fine.

Oh! thanks. I didn't noticed that. I had an issue that some apps created malformed service files. (in this case lact.service) now I know why xD

it seems like the enable is `true` by default. [systemd.services.\<name\>.enable](https://search.nixos.org/options?channel=unstable&show=systemd.services.%3Cname%3E.enable)
so this should be enough as well:
```
  environment.systemPackages = with pkgs; [                                                                                                                                            
    lact                                                                                                                                                                                                                                                                     
  ];                                                                                                                                                                                                                                                                                                                                           
  systemd.packages = with pkgs; [ lact ];
```

**Comment by ahydronous** [comment]
Shouldn't this be wrapped up in a `programs.lact.enable` so future users don't have to dig around?

**Comment by mksafavi** [comment]
> Shouldn't this be wrapped up in a `programs.lact.enable` so future users don't have to dig around?
> 

That's a good idea 👍

> I also had to use the bit of config on the wiki.
> 
> ```
>   systemd.services.lactd = {
>     description = "AMDGPU Control Daemon";
>     enable = true;  
>     serviceConfig = {
>       ExecStart = "${pkgs.lact}/bin/lact daemon";
>     };
>     wantedBy = ["multi-user.target"];
>   };
> ```
> 

I don't know why it's necessary to create this service while the service file already exists in the build. The wiki changelog doesn't explain it either.

> Neither `systemd.packages = with pkgs; [ lact ];` nor
> 
> ```
>   systemd.packages = with pkgs; [ lact ];                                                                                                                                                    
>   systemd.services.lactd.enable = true; 
> ```
> 
> worked.

What exactly didn't work?

In my config I have these two lines: There are both `lact` without d. 
```
  environment.systemPackages = with pkgs; [    lact  ];
  systemd.packages = with pkgs; [ lact ];
```

**Comment by ahydronous** [comment]
@mksafavi 

If I put your lines in my config. it won't start the daemon and LACT will start in "embedded mode" (aka doesn't do anything).
I do install LACT with `environment.systemPackages`, but to get the daemon to start I have to use the wiki config bit.

**Comment by mksafavi** [comment]
I set it up months ago so I don't actually remember if my config is enough
to auto start it. Do you get the service after adding it there?
`systemctl status lactd.service`

**Comment by ahydronous** [comment]
For `systemctl status lactd.service` I get:
```
○ lactd.service - AMDGPU Control Daemon
     Loaded: loaded (/etc/systemd/system/lactd.service; linked; preset: enabled)
     Active: inactive (dead)
```
and LACT runs in 'embedded' mode.
If I then try `sudo systemctl enable --now lactd` I get:
```
Failed to enable unit: File /etc/systemd/system/multi-user.target.wants/lactd.service: Read-only file system
```
When I try `systemctl status lactd.service` with the wiki script, I get:
```
● lactd.service - AMDGPU Control Daemon
     Loaded: loaded (/etc/systemd/system/lactd.service; enabled; preset: enable>
     Active: active (running) since Wed 2024-08-14 14:21:09 CEST; 18s ago
   Main PID: 5937 (.lact-wrapped)
         IP: 0B in, 0B out
         IO: 0B read, 0B written
      Tasks: 3 (limit: 18955)
     Memory: 2.6M (peak: 11.1M)
        CPU: 56ms
     CGroup: /system.slice/lactd.service
             └─5937 /nix/store/hvb5b9ni2c2q2m1hs2b4dxwdpbnls997-lact-0.5.4/bin/
```
and LACT works fine.

**Comment by mksafavi** [comment]
@ahydronous 
Oh. you're right. 👍
The config that I sent wasn't enough to autostart the service. I personally only start it manually so I never noticed it. 
I think overriding the config file isn't necessary. It's only missing the wantedBy argument.
Could you try this:
```
  systemd.packages = with pkgs; [ lact ];
  systemd.services.lactd = {
    enable = true;
    wantedBy = [ "multi-user.target" ];
  };
  ```

**Comment by ahydronous** [comment]
That works!
Although, since `enable` is true by default, you can get away with the oneliner of `systemd.services.lactd.wantedBy = ["multi-user.target"];`

This brings the total to make LACT work a three-liner:
```
  environment.systemPackages = with pkgs; [ lact ];
  systemd.packages = with pkgs; [ lact ];
  systemd.services.lactd.wantedBy = ["multi-user.target"];
```

Also in case anyone stumbles on this in the future (and perhaps it should even be included in `programs.lact.enable`?), `boot.kernelParams = [ "amdgpu.ppfeaturemask=0xfffd7fff" ];` is needed to unlock voltage and clock control. Alternatively `0xffffffff` can be used, although that featuremask is occasionally reported to cause flickering.

**Comment by poperigby** [comment]
I have a custom module I made: https://codeberg.org/PopeRigby/nixos/src/branch/main/modules/nixos/services/lact/default.nix

I should probably upstream it here.

**Comment by williameliasson-dev** [comment]
This worked for me.  
  
  systemd.services.lactd = {
    description = "AMDGPU Control Daemon";
    wantedBy = [ "multi-user.target" ];
    after = [ "multi-user.target" ];
    
    serviceConfig = {
      ExecStart = "${pkgs.lact}/bin/lact daemon";
      Type = "simple";
      # Run as root since we need direct hardware access
      User = "root";
      Group = "root";
      Restart = "on-failure";
      RestartSec = "5";
    };
  };

**Comment by MerkuR-92** [comment]
Since the package still hasn't been updated and version 0.6.0 is far behind the current 0.7.4 (including key features like support for Nvidia graphics), I think the best solution would be to create the service using the daemon directly from the Flatpak version (the one located at `/var/lib/flatpak/app/io.github.ilya_zlobintsev.LACT/current/*/files/bin/lact`). However, it seems this isn’t easy to do, and when using that path, the following error appears:

> NixOS cannot run dynamically linked executables intended for generic Linux environments out of the box. For more information, see: https://nix.dev/permalink/stub-ld

I believe this could be key to always having the most up-to-date version (from the official Flatpak).
